src/str/strutil.3
.\"(c) Copyright 1992, 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms 
.\"and conditions for redistribution.
.\"
.\" $Id: strutil.3,v 1.1 2001/05/26 22:04:51 seth Exp $
.TH STRUTIL 3X "30 September 1992"
.SH NAME
str_find, str_casefind, str_fill, str_lower, str_upper -- string utility functions
.SH SYNOPSIS
.LP
.nf
.ft B
#include "str.h"
.LP
.ft B
char *str_find( s1, s2 )
char *s1, *s2 ;
.LP
.ft B
char *str_casefind( s1, s2 )
char *s1, *s2 ;
.LP
.ft B
void str_fill( s, c )
char *s ;
char c ;
.LP
.ft B
char *str_lower( s )
char *s ;
.LP
.ft B
char *str_upper( s )
char *s ;
.SH DESCRIPTION
.B str_find()
returns a pointer to the first instance of string \fIs2\fR in string \fIs1\fR.
If \fIs2\fR is the empty string a pointer to \fIs1\fR is returned.
.LP
.B str_casefind()
performs the same function as
.B str_find()
except that it performs case insensitive character comparisons.
.LP
.B str_fill()
fills the string \fIs\fR with the character \fIc\fR.
.LP
.B str_lower()
and
.B str_upper()
convert their argument in place to a lower or upper case string respectively.
.SH "RETURN VALUES"
.LP
\fBstr_find()\fR 
and
.B str_casefind()
return a pointer to the first occurrence of \fIs2\fR
in \fIs1\fR or
.SM NULL
if \fIs2\fR does not exist in \fIs1\fR.
.LP
\fBstr_lower()\fR and \fBstr_upper()\fR return \fIs\fR.


ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive
	operations on empty hashes.

2002-05-13  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): Shared libfsma no longer
	breaks with libtool (on both linux and bsd).

2002-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): back out shared libfsma;
	this breaks libtool in a strange way that requires investigation

2002-05-10  James Tanis  <jtt@winwood.sysdetect.com>

	* src/fsma/fsma.c: Formatting and simplify coalesce code.

	* src/fsma/Makefile (BK_LARGE_LIB): Make libfsma shared.

	* src/dict/bst.c (bst_redelete): Improve efficiency by avoiding
	unnecessary checks and recusive calls.

2002-02-22  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_NUKE): delete contents and dict itself
	(_DICT_NUKE_WORK_LOOP): common code for DICT_NUKE, DICT_NUKE_CONTENTS

	* src/dict/dictimpl.h (HANDLE_ERROR): Eliminate unused id argument
	[use BK_FUNCNAME  here instead if you need it for debugging]

2002-01-31  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/Makefile (BK_LARGE_LIB): use ${LIBEXT} for libtool build

2002-01-09  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/fsma.c: INSIGHT->INSURE


src/str/strutil.c
/*
 * (c) Copyright 1992, 1993 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms
 * and conditions for redistribution.
 */

UNUSED static char RCSid[] = "$Id: strutil.c,v 1.2 2003/06/17 05:10:55 seth Exp $" ;


#include <ctype.h>
#include "str.h"
#include "clchack.h"

#ifndef NULL
#define NULL				0
#endif


#ifndef TRIVIAL_STR_FIND

/*
 * look for an instance of s2 in s1
 * Returns a pointer to the beginning of s2 in s1
 */
char *str_find(register char *str, register char *sstr)
{
   register int ssfc = *sstr++ ;    /* sub-string first char */

	if ( ssfc == 0 )			/* empty string is always a match */
		return( str ) ;

	while ( *str )
	{
		char *current = str ;
		register int strc = *str++ ;
		register char *sp ;                    /* string pointer */
		register char *ssp ;                   /* sub-string pointer */

		if ( strc != ssfc )
			continue ;
	
		/*
		 * We don't need to make the end of str a special case since
		 * the comparison of *sp against *ssp is guaranteed to fail
		 */
		for ( sp = str, ssp = sstr ;; sp++, ssp++ )
		{
			if ( *ssp == 0 )
				return( current ) ;
			if ( *sp != *ssp )
				break ;
		}
	}

	return( 0 ) ;
}


#define LOWER_CASE( c )					( (c) + 'a' - 'A' )

/*
 * str_casefind is similar to str_find except that it ignores the
 * case of the alphabetic characters
 */
char *str_casefind(register char *str, char *sstr)
{
	register int ssfc = *sstr++ ;		/* sub-string first char */

	if ( ssfc == 0 )
		return( str ) ;

	if ( isalpha( ssfc ) && isupper( ssfc ) )
		ssfc = LOWER_CASE( ssfc ) ;

	while ( *str )
	{
		char *current = str ;
		register int strc = *str++ ;
		char *sp ;							/* string pointer */
		char *ssp ;							/* sub-string pointer */

		if ( isalpha( strc ) && isupper( strc ) )
			strc = LOWER_CASE( strc ) ;
		if ( strc != ssfc )
			continue ;
	
		for ( sp = str, ssp = sstr ;; sp++, ssp++ )
		{
			register int sc = *sp ;				/* string char */
			register int ssc = *ssp ;			/* substring char */

			/*
			 * End-of-substring means we got a match
			 */
			if ( ssc == 0 )
				return( current ) ;

			/*
			 * Convert to lower case if alphanumeric
			 */
			if ( isalpha( sc ) && isupper( sc ) )
				sc = LOWER_CASE( sc ) ;
			if ( isalpha( ssc ) && isupper( ssc ) )
				ssc = LOWER_CASE( ssc ) ;
			if ( sc != ssc )
				break ;
		}
	}

	return( 0 ) ;
}


#else		/* defined( TRIVIAL_STR_FIND ) */

/*
 * look for an instance of s2 in s1
 * Returns a pointer to the beginning of s2 in s1
 */
char *str_find( s1, s2 )
	char *s1 ;
	char *s2 ;
{
   int i ;
   int l1 = strlen( s1 ) ;
   int l2 = strlen( s2 ) ;

   for ( i = 0 ; i < l1 - l2 + 1 ; i++ )
      if ( strncmp( &s1[ i ], s2, l2 ) == 0 )
         return( &s1[ i ] ) ;
   return( NULL ) ;
}


char *str_casefind( s1, s2 )
	char *s1 ;
	char *s2 ;
{
   int i ;
   int l1 = strlen( s1 ) ;
   int l2 = strlen( s2 ) ;

   for ( i = 0 ; i < l1 - l2 + 1 ; i++ )
      if ( strncasecmp( &s1[ i ], s2, l2 ) == 0 )
         return( &s1[ i ] ) ;
   return( NULL ) ;
}

#endif 	/* TRIVIAL_STR_FIND */


/*
 * Fill string s with character c
 */
void str_fill(register char *s, register char c)
{
	while ( *s ) *s++ = c ;
}


char *str_lower(char *s)
{
	register char *p ;
	register int offset = 'a' - 'A' ;

	for ( p = s ; *p ; p++ )
		if ( isascii( *p ) && isupper( *p ) )
			*p += offset ;
	return( s ) ;
}


char *str_upper(char *s)
{
	register char *p ;
	register int offset = 'a' - 'A' ;

	for ( p = s ; *p ; p++ )
		if ( isascii( *p ) && islower( *p ) )
			*p -= offset ;
	return( s ) ;
}


